src/api/tokenList/TokenListApi.ts

```diff
@@ -108,6 +122,55 @@ export class TokenListApiImpl extends GenericSubscriptions<TokenDetails[]> imple
     currentUserList.push(token)
     localStorage.setItem(storageKey, JSON.stringify(currentUserList))
   }
+
+  private persistTokensForNetwork(networkId: number): void {
+    const storageKey = TokenListApiImpl.getLocalStorageKey(networkId)
+    localStorage.setItem(storageKey, JSON.stringify(this._tokensByNetwork[networkId]))
+  }
+
+  private async updateTokenIds(networkId: number): Promise<void> {
+    // Set as updated on start to prevent concurrent queries
+    this.updatedIdsForNetwork.add(networkId)
+
+    // Sequentially fetch updated ids
+    const updatedTokenList = []
+    let failedToUpdate = false
+    for (let i = 0; i < this._tokensByNetwork[networkId].length; i++) {
+      const token = this._tokensByNetwork[networkId][i]
+      const tokenAddress = token.address
+
+      try {
+        token.id = await this.exchangeApi.getTokenIdByAddress({ networkId, tokenAddress })
+        updatedTokenList.push(token)
+
+        logDebug(`[network:${networkId}][address:${tokenAddress}] Token id updated: ${token.id}`)
+      } catch (e) {
+        if (e.message.match(/Must have Address to get ID/)) {
```

Comment:
Shouldn't we check by using `tokenAddressToIdMap`, sounds a bit strange this check

Reply:
Not really. In my view error handling is up to the caller.

Reply:
Yes, I know, but the strange part is how we check which kind of error it is, right?

wait wait, I misunderstood there, I thought it was handling the case where the address is not in the exchange. But it's just checking the implementation I see:
```js
public async getTokenIdByAddress({ tokenAddress }: GetTokenIdByAddressParams): Promise<number> {
    assert(typeof this.tokenAddressToId[tokenAddress] === 'number', 'Must have Address to get ID')
    return this.tokenAddressToId[tokenAddress]
  }
```

So I still think it's a bid odd. Cause it's relying on a message of an exception of an specific implementation of the API. If someone changes this message, the Api implementation or whatever, this piece won't realise. Or maybe I miss sth.

Reply:
> Cause it's relying on a message of an exception of an specific implementation of the API. If someone changes this message, the Api implementation or whatever, this piece won't realise

That is true.

What we could do is to have error codes, like what Dima did for the service response here https://github.com/gnosis/dex-react/pull/818/files#diff-e89dfdf83d68cb15aaac942477290d0bR52

But, this won't change the fact that the implementation can change and we are at the same problem as you point out.

Although, the contract implementation requires a new deployment, which doesn't happen every day. So I'd say we are safe on that front.

Reply:
Aha, I was not 100% sure where that error message came from. I see is in the IdToAddressBiMap. Yes, we don't expect this to change easily, won't happen unless we change the contract implementation. Hoewever, I find that if this "service logic" is relying 
on handling that case, we are better of using the funcion with a pink arrow, this way, we are type safe. If we change the smart contract implementation, that would give us a compile error. Also, it's more declarative.

![image](https://user-images.githubusercontent.com/2352112/78402981-da6c6780-75fb-11ea-821a-21bfee3dacb5.png)

Reply:
Aha, this function then: https://github.com/gnosis/dex-contracts/blob/master/contracts/BatchExchange.sol#L417

Will do

Reply:
https://github.com/gnosis/dex-react/pull/824/commits/ef569fce6b8a0f4bf9b920194a4e1983edb61bbd